### PR TITLE
[expat] Update the version to 2.2.9 (add support for uwp)

### LINKED ISSUE #3174: wxwidgets:x64-uwp requires expat:x64-uwp, which does not exist
I am trying to install wxwidgets:x64-uwp which does exist, however it has a dependency on expat:x64-uwp which does not exist yet.  Any thoughts on when it will be available?

Package: expat:x64-uwp
Vcpkg version: 0.0.107-2018-03-28-10e77be9fa5a4d6bb132c8d56ee8426de36ab449

Thank you.

### PR DESCRIPTION
**Pull request by NancyLi1013**
Update to 2.2.9
- Add support for uwp
- Support` find_package() `by `expat-config.cmake` instead of `Findexpat.cmake`(Remove usage in _ports/expat/usage_)

Fix #3174,  #11840

### CODE REVIEW

**Comment by ray42** [comment]
@NancyLi1013 @JackBoosY what needs to be done to get this merged? Anything I could help with? I would love to get expat v2.2.9, thanks.

**Comment by JackBoosY** [comment]
@ray42 we are trying to solve the baseline error, then we will handle this PR.

**Comment by ray42** [comment]
@JackBoosY Thank you.

@NancyLi1013 - let me know if there is anything I can help out with, thank you.

**Comment by ray42** [comment]
The same tests have been failing for quite a while. Are they expected to fail? If so, should we add them to ci.baseline.txt ?
If not, is there anything I can do to help push this along? Thank you for everything you're doing, I appreciate it a lot.

**Comment by NancyLi1013** [comment]
@ray42 
Thanks for paying attention to this PR.
I will continue solving the regressions on this PR.
I will ask for your help if there is something need to be done from you.

**Comment by oswjk** [comment]
I also just want to thank anyone working on this. Keep up the good work 👍

**Comment by NancyLi1013** [comment]
It should work now.
@ray42 
Could you please help review this PR?
Any suggestions are welcomed.

File: ports/expat/portfile.cmake [hunk 1]
```diff
@@ -20,28 +18,33 @@ vcpkg_configure_cmake(
...
+        -DEXPAT_BUILD_TESTS=OFF
+        -DEXPAT_BUILD_TOOLS=OFF
+        -DEXPAT_SHARED_LIBS=${EXPAT_LINKAGE}
 )
 
 vcpkg_install_cmake()
 
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include ${CURRENT_PACKAGES_DIR}/debug/lib/pkgconfig ${CURRENT_PACKAGES_DIR}/lib/pkgconfig)
-file(INSTALL ${SOURCE_PATH}/expat/COPYING DESTINATION ${CURRENT_PACKAGES_DIR}/share/expat RENAME copyright)
+vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/${PORT})
```
Comment (JackBoosY) [inline ports/expat/portfile.cmake:29]:
```suggestion
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/expat)
```

File: ports/gdal/portfile.cmake [hunk 2]
```diff
@@ -86,12 +80,17 @@ if (NOT VCPKG_CMAKE_SYSTEM_NAME OR VCPKG_CMAKE_SYSTEM_NAME STREQUAL "WindowsStor
...
   endif()
-
+  
   # Setup expat libraries + include path
   file(TO_NATIVE_PATH "${CURRENT_INSTALLED_DIR}/include" EXPAT_INCLUDE_DIR)
-  file(TO_NATIVE_PATH "${CURRENT_INSTALLED_DIR}/lib/expat.lib" EXPAT_LIBRARY_REL)
-  file(TO_NATIVE_PATH "${CURRENT_INSTALLED_DIR}/debug/lib/expat.lib" EXPAT_LIBRARY_DBG)
-
+  if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+      file(TO_NATIVE_PATH "${CURRENT_INSTALLED_DIR}/lib/libexpatMD.lib" EXPAT_LIBRARY_REL)
```
Comment (JackBoosY) [inline ports/gdal/portfile.cmake:87]: We should consider the non-Windows situations.

File: ports/readosm/portfile.cmake [hunk 3]
```diff
@@ -16,13 +16,14 @@ vcpkg_extract_source_archive_ex(
 
 find_program(NMAKE nmake)
 
-set(LIBS_ALL_DBG "\"${CURRENT_INSTALLED_DIR}/debug/lib/expat.lib\" \"${CURRENT_INSTALLED_DIR}/debug/lib/zlibd.lib\"")
-set(LIBS_ALL_REL "\"${CURRENT_INSTALLED_DIR}/lib/expat.lib\" \"${CURRENT_INSTALLED_DIR}/lib/zlib.lib\"")
-
 if(VCPKG_CRT_LINKAGE STREQUAL dynamic)
+    set(LIBS_ALL_DBG "\"${CURRENT_INSTALLED_DIR}/debug/lib/libexpatd.lib\" \"${CURRENT_INSTALLED_DIR}/debug/lib/zlibd.lib\"")
```
Comment (JackBoosY) [inline ports/readosm/portfile.cmake:20]: Same as above.

File: ports/spatialite-tools/portfile.cmake [hunk 4]
```diff
@@ -30,6 +28,8 @@ if(VCPKG_CRT_LINKAGE STREQUAL dynamic)
     set(SPATIALITE_LIBS_DBG "${LDIR}/debug/lib/spatialite.lib")
     set(ICONV_LIBS_REL "${LDIR}/lib/libiconv.lib")
     set(ICONV_LIBS_DBG "${LDIR}/debug/lib/libiconv.lib")
+    set(EXPAT_LIBS_REL "${LDIR}/lib/libexpat.lib")
```
Comment (JackBoosY) [inline ports/spatialite-tools/portfile.cmake:31]: Same as above.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by strega-nil** [comment]
This seems like it should be merged; should we resolve conflicts and do so @NancyLi1013 ?

**Comment by ray42** [comment]
Thank you @NancyLi1013 for all your hard work. Does it look like this can be merged now?

**Comment by NancyLi1013** [comment]
@ray42 
Thanks for your kindly reminder.
It should be fine on my sides. But since these features are blocked by other ports. 
 I'm not sure if all features can work well now.

**Review by JackBoosY — APPROVED** [review]

**Comment by NancyLi1013** [comment]
All features have finished testing.

- `skia[metal] `failed on x64-windows and x64-windows-static. (Also failed on master branch)

- `poco[mariadb,postgresql,pdf] `have passed on x64-windows.
- `poco[mysql] ` failed on x64-windows. (Also failed on master branch)
- `poco[mariadb,postgresql]` failed on x64-windows-static. (Also failed on master branch)
- `poco[pdf]` has passed on x64-windows-static
- `exiv2[unicode,xmp]` have passed on x86-windows and x64-windows-static.
- `gdal[mysql-libmariadb,libspatialite]` have passed on x64-windows.
- `gdal[mysql-libmariadb,libspatialite]` failed on x64-windows-static. (Also failed on master branch)

**Comment by NancyLi1013** [comment]
The regression on x64-osx for `itk `is due to the following cause, which will be fixed in PR #11208.
```
Undefined symbols for architecture x86_64:
  "gdcm::Rescaler::SetMinMaxForPixelType(double, double)", referenced from:
      itk::GDCMImageIO::Write(void const*) in libITKIOGDCM-5.0.a(itkGDCMImageIO.cxx.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

**Comment by strega-nil** [comment]
@NancyLi1013 is this `itk` failure going to happen regardless, or should we wait for #11208 to be merged?

**Comment by NancyLi1013** [comment]
The `itk `failure will happen regardless, which has existed for long time . It seems that this issue appeared in this PR https://github.com/microsoft/vcpkg/pull/10767#issuecomment-629959205 for the first time and the failure code was removed from ci.baseline.txt in #11513. It appeared again in #https://github.com/microsoft/vcpkg/pull/12067#issuecomment-649738044 and https://github.com/microsoft/vcpkg/pull/12133.

It should not block this PR. But we can also wait for #11208 merged and then try to rerun this PR. There are also many changes in this PR. We may wait for all CI checks have passed.

**Comment by strega-nil** [comment]
@NancyLi1013 what's going on with `itk` and `cmake` on windows, do you know?

**Comment by NancyLi1013** [comment]
@strega-nil 

For the regressions caused by itk, I'm in investigation now.
For cmake, I don't know why it will fail.

I resolved the conflicts for `wxwidgets` just now. Let's wait for the new results.

**Comment by NancyLi1013** [comment]
The regressions for `itk `have been fixed now. 
But for `cmake`, I have no idea.

```
[503/504] cmd.exe /C "cd /D D:\buildtrees\cmake\x64-windows-dbg && D:\buildtrees\cmake\x64-windows-dbg\bin\cmake.exe -P cmake_install.cmake"
FAILED: CMakeFiles/install.util 
cmd.exe /C "cd /D D:\buildtrees\cmake\x64-windows-dbg && D:\buildtrees\cmake\x64-windows-dbg\bin\cmake.exe -P cmake_install.cmake"
ninja: build stopped: subcommand failed.
```

**Comment by NancyLi1013** [comment]
@strega-nil 
Thanks for your help to fix cmake issue.

I noticed that you updated cmake to 3.18.1 and also disable `-DCMAKE_USE_SYSTEM_EXPAT=OFF`.

I tried to install cmake 3.17.2 via  `.\vcpkg.exe install cmake --overlay-ports=scripts/test_ports `on my local on both master branch and expat branch, it failed with the same error, which is different from CI results.

The results from expat branch:
```
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(39): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(39): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(55): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(55): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(57): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(57): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(217): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(217): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\expet\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
```
The results from master branch:
```
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(39): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(39): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(55): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(55): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(57): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(57): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(217): error C2672: 'cm::static_reference_cast': no matching overloaded function found
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Source\cmGeneratorExpressionParser.cxx(217): error C2783: 'T &cm::static_reference_cast(O &)': could not deduce template argument for '__formal'
F:\new\vcpkg\buildtrees\cmake\src\7fc4616c28-392dff6306.clean\Utilities\std\cmext/memory(19): note: see declaration of 'cm::static_reference_cast'
```
So I'm not sure if the regressions on CI are related with the changes in this PR. I also have a doubt that why you update cmake to 3.18.1?

**Comment by strega-nil** [comment]
@NancyLi1013 yeah, that was a thing that was broken in that version of CMake with the latest version of VC++.

**Comment by NancyLi1013** [comment]
@strega-nil thanks for your help.

**Comment by ray42** [comment]
Thanks to everyone, especially @NancyLi1013 , for the awesome work :)

**Pull request merged**